### PR TITLE
OntologyManager: add reason() + test

### PR DESCRIPTION
**Pull request by Tadinu**
Following #141, this PRs add reasoning APIs to `OntologyManager`, specifically to allow dynamic SWRL rules to be created for instances of an ontology concept class.
* Add `reason()`: reason on ontology concepts and their relations of a specific ontology world. Calling this will generate deducted facts that are updated to contents of property instances.
* Basic rules of binary relation (reflexivity, symmetry, transitivity) are introduced. NOTE: These just act as basic examples of rule definitions and more emphasize on the dynamic creation aspect. Same effects could be also achieved but incur creating new property classes, subclassing `ReflexiveProperty, SymmetricProperty, TransitiveProperty`
* Fix a bug causing reasoning to fail, due to missing required specific world as domain of ontology concepts for reasoning.
-> `LoadOntology()` now by default load from an IRI to the owlready2-provided `default_world`, which is also by default used for 
`sync_reasoner_pellet()/sync_reasoner_hermit()`.
-> Alternatively, `create_ontology_world()` is added in case of loading ontologies into a custom persistent world & reasoning on it specifically.
* `test_ontology.py`: Add `test_reasoning()`, also do post-cleaning up of `owl/sql, sql-journal` files generated during the test.
* `create_ontology_concept_class()/create_ontology_property_class()`: add optional `ontology` param to allow creating them into a specific ontology namespace, aside from the default `main_ontology`.
* `create_ontology_triple_classes()`:
  + (1) `predicate_name -> predicate_class_name`, `inverse_predicate_name -> inverse_predicate_class_name` for clarity
  + (2) add `predicate_python_attribute_name, inverse_predicate_python_attribute_name`, by default taking values from (1) if not  explicitly specified, for consistency
  + Make `ontology_object_parent_class` type hint as `Thing | object` to support such an `owlready2.FunctionalProperty` relation between `owlready2.Thing` and pure python types
* Docker/CI: add & verify Java runtime installation as required by the reasoner
https://owlready2.readthedocs.io/en/latest/reasoning.html#configuration

### CODE REVIEW

**Comment by Tadinu** [comment]
test_ontology_reasoning() currently fails due to JRE, which is required by owlready2 reasoner, is not available atm

**Comment by Tadinu** [comment]
test_ontology_reasoning() currently fails since JRE, which is required by owlready2 reasoner, is not available atm

File: test/test_ontology.py [hunk 1]
```diff
@@ -38,35 +41,78 @@ def setUpClass(cls):
...
+        extra_sql_world_2 = self.ontology_manager.create_ontology_world(use_global_default_world=False,
+                                                                        sql_backend_filename=extra_world_sql_filename)
+        self.assertIsNotNone(extra_sql_world_2)
+
+        # Remove SQL file finally
+        self.remove_sql_file(extra_world_sql_filename)
 
     def test_ontology_concept_holder(self):
-        if not self.ontology_manager.initialized():
+        if not owlready2:
```
Comment (Tigul) [inline test/test_ontology.py:94]: There is a @skipif that should be used in this case

File: test/test_ontology.py [hunk 2]
```diff
@@ -38,35 +41,78 @@ def setUpClass(cls):
...
-        dynamic_ontology_concept_holder = OntologyConceptHolder(dynamic_ontology_concept_class(name='dynamic_ontology_concept1',
-                                                                                               namespace=self.main_ontology))
+        dynamic_ontology_concept_holder = OntologyConceptHolder(
+            dynamic_ontology_concept_class(name='dynamic_ontology_concept1',
+                                           namespace=self.main_ontology))
         self.assertTrue(owlready2.isinstance_python(dynamic_ontology_concept_holder.ontology_concept, owlready2.Thing))
 
     def test_loaded_ontologies(self):
-        if not self.ontology_manager.initialized():
+        if not owlready2:
```
Comment (Tigul) [inline test/test_ontology.py:103]: @skipif

File: test/test_ontology.py [hunk 3]
```diff
@@ -38,35 +41,78 @@ def setUpClass(cls):
...
+        if self.ontology_manager.main_ontology_iri is SOMA_ONTOLOGY_IRI or \
+                self.ontology_manager.main_ontology_iri is SOMA_HOME_ONTOLOGY_IRI:
+            self.assertIsNotNone(self.soma)
+            self.assertTrue(self.soma.loaded)
+            self.assertIsNotNone(self.dul)
+            self.assertTrue(self.dul.loaded)
 
     def test_ontology_concept_class_dynamic_creation(self):
-        if not self.ontology_manager.initialized():
+        if not owlready2:
```
Comment (Tigul) [inline test/test_ontology.py:115]: @skipif

File: test/test_ontology.py [hunk 4]
```diff
@@ -78,13 +124,13 @@ def test_ontology_concept_class_dynamic_creation(self):
         self.assertTrue(owlready2.isinstance_python(dynamic_ontology_concept, owlready2.Thing))
 
     def test_ontology_triple_classes_dynamic_creation(self):
-        if not self.ontology_manager.initialized():
+        if not owlready2:
```
Comment (Tigul) [inline test/test_ontology.py:127]: @skipif

File: src/pycram/ontology/ontology.py [hunk 5]
```diff
@@ -84,24 +98,117 @@ def print_ontology_class(ontology_class: Type[Thing]):
...
+            # Create a new world with parallelized file parsing enabled
+            if sql_backend_path_valid:
+                world = OntologyWorld(filename=sql_backend_filename, exclusive=False, enable_thread_parallelism=True)
+            else:
+                world = OntologyWorld(exclusive=False, enable_thread_parallelism=True)
+            rospy.loginfo(f"Created a new ontology world with SQL backend: {sql_backend_name}")
+        return world
 
-    def load_ontology(self, ontology_iri: str) -> tuple[Ontology, Namespace]:
+    def load_ontology(self, ontology_iri: str) -> tuple[Optional[Ontology], Optional[Namespace]]:
```
Comment (Tigul) [inline src/pycram/ontology/ontology.py:190]: From what I can see the return should be Optional[Tuple[Ontology, Namespace]]

**Review by Tigul — CHANGES_REQUESTED** [review]
Please change

**Review by Tigul — APPROVED** [review]

**Pull request merged**